internal/locate/store_cache.go

```diff
@@ -1098,3 +1117,90 @@ func (s *Store) resetReplicaFlowsStats(destType replicaFlowsType) {
 func (s *Store) recordReplicaFlowsStats(destType replicaFlowsType) {
 	atomic.AddUint64(&s.replicaFlowsStats[destType], 1)
 }
+
+func updateStoreLivenessGauge(store *Store) {
+	if store.storeType != tikvrpc.TiKV || store.getResolveState() != resolved {
+		return
+	}
+	metrics.TiKVStoreLivenessGauge.WithLabelValues(strconv.FormatUint(store.storeID, 10)).Set(float64(store.getLivenessState()))
+}
+
+type storeCacheUpdater struct {
+	stores storeCache
+
+	lastCleanUpTime  time.Time
+	nextCleanUpStore uint64
+}
+
+func (u *storeCacheUpdater) tick(ctx context.Context, now time.Time) bool {
+	storeList, err := u.stores.fetchAllStores(ctx, opt.WithExcludeTombstone())
+	if err != nil {
+		logutil.Logger(ctx).Info("refresh full store list failed", zap.Error(err))
+		return false
+	}
+	u.insertMissingStores(ctx, storeList)
+	u.cleanUpStaleStoreMetrics(ctx, storeList, now)
+	return false
+}
+
+// insertMissingStores adds the stores that are not in the cache.
+func (u *storeCacheUpdater) insertMissingStores(ctx context.Context, storeList []*metapb.Store) {
+	for _, store := range storeList {
+		// storeList is supposed to contains only Up and Offline stores.
+		// This check is being defensive and to make it consistent with store resolve code.
+		if store == nil || store.GetState() == metapb.StoreState_Tombstone {
+			continue
+		}
+		_, exist := u.stores.get(store.GetId())
+		if exist {
+			continue
+		}
+		s := u.stores.getOrInsertDefault(store.GetId())
+		if err := s.initResolveLite(store); err != nil {
+			logutil.Logger(ctx).Warn("init resolve store failed", zap.Uint64("storeID", store.GetId()), zap.Error(err))
+			continue
+		}
+		updateStoreLivenessGauge(s)
+	}
+}
+
+func (u *storeCacheUpdater) cleanUpStaleStoreMetrics(ctx context.Context, storeList []*metapb.Store, now time.Time) {
+	if now.Sub(u.lastCleanUpTime) < cleanStoreMetricsInterval {
+		return
+	}
+	u.lastCleanUpTime = now
+
+	// find a stale store id
+	id := u.nextCleanUpStore
+	if id == 0 {
+		validStoreIDs := make(map[uint64]struct{}, len(storeList))
+		for _, store := range storeList {
+			if store.GetId() != 0 && store.GetState() != metapb.StoreState_Tombstone {
+				validStoreIDs[store.GetId()] = struct{}{}
+			}
+		}
+		u.nextCleanUpStore = metrics.FindNextStaleStoreID(metrics.TiKVStoreLivenessGauge, validStoreIDs)
+		// delay cleaning up this store for one cleanStoreMetricsInterval
+		return
+	} else {
+		u.nextCleanUpStore = 0
+	}
+
+	// confirm that the store is indeed stale from pd
+	store, err := u.stores.fetchStore(ctx, id)
+	if err != nil {
```

Comment:
Is it necessary to check `isStoreNotFoundError` here?

Reply:
It would be better to check that to avoid some data residue. Added.